### PR TITLE
Rework upsert individual function

### PR DESCRIPTION
**Pull request by davereinhart**
To prevent gaps in the `individual_id` sequence and unnecessary updates to records
in `warehouse.individual` when there are no changes, the `upsert_individual` function
is being updated to first check if a record exists and then only perform the update
if the relevant columns have different data. Preventing unnecessary updates will help
reduce bloat and provide a more accurate modified date on the record.

### CODE REVIEW

**Comment by davereinhart** [comment]
While testing this function, I discovered that a previous update of this function resulted in some extraneous null values in `individual.details` records coming. That issue is resolved with this update, but will require cleaning up those records on prod after deployment:
```
--update warehouse.individual set details = null
select * from warehouse.individual
where jsonb_strip_nulls(individual.details->0) = '{}'::jsonb
```

File: lib/id3c/cli/command/etl/__init__.py [hunk 1]
```diff
@@ -78,23 +78,64 @@ def upsert_individual(db: DatabaseSession, identifier: str, sex: str = None, det
...
+                returning individual_id as id, identifier
+        """, { **data, "individual_id": individual.id })
+
+    # More than one found → error
+    else:
+        raise Exception(f"More than one individual found matching identifier: {identifier}")
 
-    LOG.info(f"Upserted individual {individual.id} «{individual.identifier}»")
+    if individual:
+        LOG.info(f"Upserted encounter {individual.id} «{identifier}»")
```
Comment (bencap) [inline lib/id3c/cli/command/etl/__init__.py:138]:
```suggestion
        LOG.info(f"Upserted individual {individual.id} «{identifier}»")
```

File: lib/id3c/cli/command/etl/__init__.py [hunk 2]
```diff
@@ -78,23 +78,64 @@ def upsert_individual(db: DatabaseSession, identifier: str, sex: str = None, det
...
-        insert into warehouse.individual (identifier, sex, details)
-            values (%(identifier)s, %(sex)s, %(details)s)
+    with db.cursor() as cursor:
+        cursor.execute("""
+            select individual_id as id,
+                identifier,
+                (
+                    individual.sex != %(sex)s
+                    or
+                    (coalesce(individual.details,'{}'::jsonb) != (coalesce(individual.details, '{}'::jsonb) || coalesce(%(details)s, '{}'::jsonb)))
```
Comment (bencap) [inline lib/id3c/cli/command/etl/__init__.py:91]: Curious about this line, in what circumstances would `(coalesce(individual.details,'{}'::jsonb) != (coalesce(individual.details, '{}'::jsonb)` these two things not be equal to each other? Does it have something to do with the null values you noted in your comment above?

**Review by bencap — COMMENTED** [review]
Looked good to me based on local testing, just a minor typo in one of the log statements.

File: lib/id3c/cli/command/etl/__init__.py (hunk 2 above)
Comment (davereinhart) [reply lib/id3c/cli/command/etl/__init__.py:91]: The two values being compared there are:

`coalesce(individual.details,'{}'::jsonb)` 
and
`(coalesce(individual.details, '{}'::jsonb) || coalesce(%(details)s, '{}'::jsonb))`

The latter is the merged value that would be used if an update is performed. And yes, the coalesces are there to make sure it handles nulls correctly. The current version is resulting in jsonb arrays where a null is appended each time there's an update but this change should prevent that.

Comment (bencap) [reply lib/id3c/cli/command/etl/__init__.py:91]: Ah got it thanks, I was misreading the statement.

**Review by bencap — APPROVED** [review]
Looked good to me based on local testing, just a minor typo in one of the log statements.

**Review by kschwabefry — APPROVED** [review]

**Comment by davereinhart** [comment]
Deployed.

**Pull request merged**